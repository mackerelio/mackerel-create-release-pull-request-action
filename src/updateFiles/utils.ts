import fs from "fs";
import { create } from "@actions/glob";

// Rewrite the contents of the matching file using callback
export async function replaceContents(
  argGlob: readonly string[],
  callback: (content: string, path: string) => string
): Promise<void> {
  const globber = await create(argGlob.join("\n"));
  const paths = await globber.glob();

  for (const path of paths) {
    let content = callback(await fs.promises.readFile(path, "utf8"), path);

    // Add If the new line at the end of the content does not exist
    if (!/[\r\n]$/.test(content)) {
      content += "\n";
    }

    // Inherit permissions.
    const stat = await fs.promises.stat(path);
    await fs.promises.writeFile(path, content, { encoding: "utf8", mode: stat.mode });
  }
}
