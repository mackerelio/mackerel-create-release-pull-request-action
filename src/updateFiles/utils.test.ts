import fs from "fs";
import { cp } from "@actions/io";
import snapshotDiff from "snapshot-diff";
import { file } from "tmp-promise";
import { replaceContents } from "./utils";

describe("replaceContents", () => {
  describe("glob", () => {
    it("dirglob", async () => {
      const files: string[] = [];
      await replaceContents(["testdata/replace/glob/*/file1"], (data, path) => {
        files.push(path);
        return data;
      });

      expect(files).toEqual([
        `${process.cwd()}/testdata/replace/glob/dir1/file1`,
        `${process.cwd()}/testdata/replace/glob/dir2/file1`,
      ]);
    });

    it("fileglob", async () => {
      const files: string[] = [];
      await replaceContents(["testdata/replace/glob/dir1/file*"], (data, path) => {
        files.push(path);
        return data;
      });

      expect(files).toEqual([
        `${process.cwd()}/testdata/replace/glob/dir1/file1`,
        `${process.cwd()}/testdata/replace/glob/dir1/file2`,
      ]);
    });
  });

  describe("can replace content", () => {
    it("keep permission", async () => {
      const { path: testFilePath } = await file();
      await fs.promises.writeFile(testFilePath, "\n", { mode: 0o755 });
      const before = await fs.promises.stat(testFilePath);

      await replaceContents([testFilePath], data => data);

      const after = await fs.promises.stat(testFilePath);
      expect(before.mode).toEqual(after.mode);
    });

    it("keep utf8 encoding", async () => {
      const initialFile = "testdata/replace/utf8.txt";
      const initial = await fs.promises.readFile(initialFile, "utf8");

      const { path: testFilePath } = await file();
      await cp(initialFile, testFilePath);

      await replaceContents([testFilePath], data => data);

      const processed = await fs.promises.readFile(testFilePath, "utf8");
      expect(initial).toEqual(processed);
    });

    it("results is ended newline", async () => {
      const initialFile = "testdata/replace/test.txt";
      const initial = await fs.promises.readFile(initialFile, "utf8");

      const { path: testFilePath } = await file();
      await cp(initialFile, testFilePath);

      await replaceContents([testFilePath], data => `${data}ddd\n`);

      const processed = await fs.promises.readFile(testFilePath, "utf8");
      expect(snapshotDiff(initial, processed)).toMatchSnapshot();
    });

    it("append newline when ended newline is not exist", async () => {
      const initialFile = "testdata/replace/test.txt";
      const initial = await fs.promises.readFile(initialFile, "utf8");

      const { path: testFilePath } = await file();
      await cp(initialFile, testFilePath);

      await replaceContents([testFilePath], data => `${data}ddd`);

      const processed = await fs.promises.readFile(testFilePath, "utf8");
      expect(snapshotDiff(initial, processed)).toMatchSnapshot();
    });

    it("replace version in Makefile", async () => {
      const initialFile = "testdata/replace/Makefile.txt";
      const initial = await fs.promises.readFile(initialFile, "utf8");

      const { path: testFilePath } = await file();
      await cp(initialFile, testFilePath);

      const nextVersion = "4.5.6";
      await replaceContents([testFilePath], data =>
        data.replace(/^VERSION( *:*= *).*?\n/ms, (match, p1) => `VERSION${p1}${nextVersion}\n`)
      );

      const processed = await fs.promises.readFile(testFilePath, "utf8");
      expect(snapshotDiff(initial, processed)).toMatchSnapshot();
    });

    it("replace fragment rpm specfile", async () => {
      const initialFile = "testdata/replace/rpmspec.txt";
      const initial = await fs.promises.readFile(initialFile, "utf8");

      const { path: testFilePath } = await file();
      await cp(initialFile, testFilePath);

      const appendChangelog = [
        "* Thu Feb 18 2021 <foo@example.com> - 1.2.4",
        "- hoge huga piyoe",
        "",
      ].join("\n");
      await replaceContents([testFilePath], data =>
        data.replace(/%changelog/, `%changelog\n${appendChangelog}`)
      );

      const processed = await fs.promises.readFile(testFilePath, "utf8");
      expect(snapshotDiff(initial, processed)).toMatchSnapshot();
    });
  });
});
