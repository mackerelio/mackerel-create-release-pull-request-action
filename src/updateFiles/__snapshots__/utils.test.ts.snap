// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceContents can replace content append newline when ended newline is not exist 1`] = `
"Snapshot Diff:
- First value
+ Second value

  aaa
  bbb
  ccc
+ ddd
"
`;

exports[`replaceContents can replace content replace fragment rpm specfile 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -3,9 +3,12 @@
  %files
  %defattr(-,root,root)
  %{_localbindir}/%{name}

  %changelog
+ * Thu Feb 18 2021 <foo@example.com> - 1.2.4
+ - hoge huga piyoe
+
  * Tue Feb 16 2021 <foo@example.com> - 1.2.3
  - foo bar baz qux

"
`;

exports[`replaceContents can replace content replace version in Makefile 1`] = `
"Snapshot Diff:
- First value
+ Second value

  BIN := mkr
- VERSION := 1.2.3
+ VERSION := 4.5.6
  CURRENT_REVISION := $(shell git rev-parse --short HEAD)
  BUILD_LDFLAGS := \\"-w -s -X main.gitcommit=$(CURRENT_REVISION)\\"
"
`;

exports[`replaceContents can replace content results is ended newline 1`] = `
"Snapshot Diff:
- First value
+ Second value

  aaa
  bbb
  ccc
+ ddd
"
`;
